### ISSUE TITLE
FileSystemNotFoundException when running from jar

### ISSUE THREAD

**Issue by jperedadnr** [FXyz/FXyz#108]
Reported from: https://stackoverflow.com/questions/62611044/filesystemnotfoundexception-while-running-a-javafx-app-on-terminal

After this [refactoring](https://github.com/FXyz/FXyz/commit/a8e519acc06328d8d1f55675df8cb9f8ff94cfc3#diff-a5ab33f08c3367762325fe178b0af5c3), when an obj model is bundled within a jar and `Importer3D.loadAsPoly(getClass().getResource(".."))` is called, there is this exception:

```
Exception in thread "JavaFX Application Thread" java.nio.file.FileSystemNotFoundException
    at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getFileSystem(ZipFileSystemProvider.java:172)
    at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getPath(ZipFileSystemProvider.java:158)
    at java.base/java.nio.file.Path.of(Path.java:208)
    at java.base/java.nio.file.Paths.get(Paths.java:98)
    at org.fxyz3d.importers.obj.ObjImporter.read(ObjImporter.java:115)
    at org.fxyz3d.importers.obj.ObjImporter.loadAsPoly(ObjImporter.java:102)
```

Similar issue happens using jlink, up until Java 13.

Proposal: Add a FileSystem if it doesn't exist.

**Comment by tkslaw** [comment]
>Proposal: Add a FileSystem if it doesn't exist.

I see two potential problems with that:

1. When embedded in a JAR the `FileSystem` implementation will likely come from the _ZIP FileSystemProvider_. This provider is typically visible to the entire application which means opening a `FileSystem` has an extremely visible side-effect.; especially since only one `FileSystem` can be open per ZIP/JAR file at any given time.

2. What if the URL's protocol has no matching `FileSystemProvider`? A simple example would be an `http[s]` URL. Granted, this may not be a common use case but the fact the API accepts a `URL` implies such a possibility.

In my opinion it would be better to revert back to using `URL#openStream()` and `BufferedReader` (at least as a fallback). Two potential counterpoints could be:

1. Using `Files#lines(Path)` returns a `Stream` which makes parallel processing easier.
    * Note that `BufferedReader#lines()` also returns a `Stream`.
2. The `Stream` returned by `Files#lines(Path)` uses an optimized `Spliterator`.
    * That's only true if (as one of three conditions) the `Path` belongs to the default `FileSystem`.

Is there some other reason to not use `URL#openStream()` and `BufferedReader` of which I'm not aware?

**Comment by jperedadnr** [comment]
Good points, pinging @nlisker...

**Comment by nlisker** [comment]
I was not aware of this limitation. Doesn't that mean that all applications that use a file system and bundled into a jar will fail? Seems like a wide scope.

We can change to the `BufferedReader` solution.

**Comment by Lorenzo-Mazza** [comment]
as I wrote on StackOverFlow, a colleague of mine suggested this fix: adding these lines at the loadAsPoly function from ObjImporter Class. 
`    public Model3D loadAsPoly(URL url) throws IOException {
        // Additional lines
             if(url.getProtocol().equals("jar")) {
                            try {
                                Map<String, String> env = new HashMap<>();
                                env.put("create", "true");
                                FileSystem zipfs = FileSystems.newFileSystem(url.toURI(), env);
                            } catch(FileSystemAlreadyExistsException ignored) {

                            } catch (IOException | URISyntaxException e) {
                                e.printStackTrace();
                            }
                        }
        // End of the addition
        return read(url, true);
    }`
I tested the Jar and it's working properly now.

**Comment by tkslaw** [comment]
@nlisker 

> I was not aware of this limitation. Doesn't that mean that all applications that use a file system and bundled into a jar will fail?

I'm not sure I understand what you mean by "use a file system and bundled into a jar". However, applications should not use the NIO.2 API to access resources. They _can_, if they open an appropriate `FileSystem`, but that's often more trouble than it's worth.

Note the `Paths#get(URI)` and `Path#of(URI)` methods ultimately delegate to [`FileSystemProvider#getPath(URI)`][3], if an appropriate provider is found. That method is documented to allow implementations to automatically create a `FileSystem` for the `URI` if one does not already exist. But it appears the `ZipFileSystemProvider` implementation does not do this and instead throws a `FileSystemNotFoundException` in such cases.

  [3]: https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/nio/file/spi/FileSystemProvider.html#getPath(java.net.URI)

---

I also want to update something I said, which was:

>especially since only one FileSystem can be open per ZIP/JAR file at any given time.

Apparently that's not 100% accurate. The [`FileSystemProvider#newFileSystem(URI,Map)`][1] method is documented to throw `FileSystemAlreadyExistsException` under obvious circumstances. But the [`FileSystemProvider#newFileSystem(Path,Map)`][2] method does not have the same restrictions. When I tried it out I was indeed able to open two `FileSystem` instances to the same ZIP file simultaneously. Unfortunately, instances created by this second factory method are not intended to be cached by the provider (i.e. there's no corresponding `#getFileSystem(Path)`).

  [1]: https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/nio/file/spi/FileSystemProvider.html#newFileSystem(java.net.URI,java.util.Map)
  [2]: https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/nio/file/spi/FileSystemProvider.html#newFileSystem(java.nio.file.Path,java.util.Map)

**Comment by nlisker** [comment]
What if we do something like this:

```
    private ObjModel read(URL url, boolean asPolygon) {
        log("Reading from URL: " + url + " as polygon: " + asPolygon);

        ObjModel model = asPolygon ? new PolyObjModel(url) : new ObjModel(url);

        try (Stream<String> lines = getLines(url)) {
            lines.map(String::trim)
                .filter(l -> !l.isEmpty() && !l.startsWith("#"))
	        .forEach(model::parseLine);
        } catch (IOException | URISyntaxException e) {
            e.printStackTrace();
        }

        ...
    }

    private Stream<String> getLines(URL url) throws IOException, URISyntaxException {
        Path path;
        try {
    	    path = Paths.get(url.toURI());
    	} catch (FileSystemNotFoundException e) {
    	    return new BufferedReader(new InputStreamReader(url.openStream())).lines();
    	}
	return Files.lines(path);
    }
```

We can replace the check for the existence of the file system in the try-catch block with the suggested `if(url.getProtocol().equals("jar"))` if it's preferable.

Thoughts?

**Comment by tkslaw** [comment]
Personally, I think the implementation shouldn't use `Files#lines(Path)` at all. I could understand using that method when there was a possibility of executing the `Stream` in parallel<sup>***</sup> but that code was removed; the `Stream` is always executed sequentially. With that in mind, and since the API only accepts `URL`s, I think the implementation should exclusively rely on the `URL` to access the contents of the object file:

```java
private ObjModel read(URL url, boolean asPolygon) {
    log("Reading from URL: " + url + " as polygon: " + asPolygon);

    ObjModel model = asPolygon ? new PolyObjModel(url) : new ObjModel(url);

    try (BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()))) {
        reader.lines()
             .map(String::trim)
             .filter(l -> !l.isEmpty() && !l.startsWith("#"))
             .forEach(model::parseLine);
    } catch (IOException e) {
        e.printStackTrace();
    }

    ...
}
```

Otherwise, note that your implementation currently has a resource leak. Closing the `Stream` from [`BufferedReader#lines()`][2]  does not close the `BufferedReader` (which I find strange) and thus leaves the `InputStream` returned by `URL#openStream()` open. This can be fixed with a call to `Stream#onClose(Runnable)`:

```java
BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()));
return reader.onClose(() -> {
    try {
        reader.close();
    } catch (IOException ex) {
        ex.printStackTrace(); // consistent with other code in this class
    }
});
```

---

<sup>*** `Files#lines(Path)` tries to use an optimized `Spliterator` when the `Path` belongs to the default `FileSystem` (see [JDK-8072773][1]). This `Spliterator` memory maps the file and shares the buffer with sub-`Spliterator`s. This makes splitting much more efficient. In contrast, `BufferedReader#lines()` uses the worst-case `Spliterator`—one entirely based on an `Iterator`. _But this benefit is only possible for `Path`s which belong to the default `FileSystem` and really only improves the performance of parallel streams._</sup>

<sup>Note that `Files#lines(Path)` resorts to `BufferedReader#lines()` when it's not possible to use the optimized `Spliterator`.</sup>

  [1]: https://bugs.openjdk.java.net/browse/JDK-8072773
  [2]: https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/io/BufferedReader.html#lines()

**Comment by jperedadnr** [comment]
@nlisker I'm reopening this issue again, I've just noticed that `MtlReader` still uses `Files.lines(Paths.get(new URI(fileUrl)))`.

Can you provide the same fix you did for `ObjImporter`?

**Comment by nlisker** [comment]
I'll have a look, yes.

**Comment by jperedadnr** [comment]
Fixed with #112
